### ISSUE TITLE
MUB Partial Writes cause mismatches in column lengths

### ISSUE THREAD

**Issue by tustvold** [influxdata/influxdb_iox#2090]
The MUB currently performs validation up-front and then assumes the actual write to be infallible.

However, #2089 lead to validated writes failing. This in turn lead to columns in the MUB containing inconsistent numbers of rows.

```
thread 'IOx Query Executor Thread' panicked at 'called `Result::unwrap()` on an `Err` value: ArrowError { source: InvalidArgumentError("all columns in a record batch must have the same length") }', mutable_buffer/src/chunk/snapshot.rs:50:14
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:92:14
   2: core::result::unwrap_failed
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/result.rs:1355:5
   3: core::result::Result<T,E>::unwrap
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/result.rs:1037:23
   4: mutable_buffer::chunk::snapshot::ChunkSnapshot::new
             at ./home/rust/project/mutable_buffer/src/chunk/snapshot.rs:47:21
   5: mutable_buffer::chunk::MBChunk::snapshot
             at ./home/rust/project/mutable_buffer/src/chunk.rs:162:33
   6: server::db::chunk::DbChunk::snapshot
             at ./home/rust/project/server/src/db/chunk.rs:110:32
```

I can see two ways to address this:

* Make the write path fallible and back out changes on error
* Validate capacity up front in addition to schema

Dictionary encoding makes it non-trivial to ascertain the necessary capacity up-front, and so the first of these is possibly the better option.

**Comment by alamb** [comment]
I agree the first option is probably best

**Comment by tustvold** [comment]
Closed by #2973
